crates/rethnet_evm/benches/util.rs

```diff
@@ -1,146 +1,246 @@
+                            state
+                                .set_account_storage_slot(
+                                    address,
+                                    U256::from(storage_slot),
+                                    U256::from(account_number),
+                                )
+                                .unwrap();
+                        }
                     }
+                    state.checkpoint().unwrap();
+                    checkpoints.push(state.state_root().unwrap());
                 }
-                state.checkpoint().unwrap();
+                snapshots.push(state.make_snapshot());
             }
         }
     }
 
     /// Returns a set of factories, each member of which produces a clone of one of the state objects in this struct.
-    fn make_clone_factories(
+    fn make_state_refs(
         &self,
     ) -> Vec<(
         &'static str, // label of the type of state produced by this factory
         Box<dyn Fn() -> Box<dyn SyncState<StateError>> + '_>,
+        &Vec<B256>,
+        &Vec<B256>,
     )> {
         vec![
-            ("LayeredState", Box::new(|| Box::new(self.layered.clone()))),
-            ("HybridState", Box::new(|| Box::new(self.hybrid.clone()))),
+            (
+                "Layered",
+                Box::new(|| Box::new(self.layered.clone())),
+                &self.layered_checkpoints,
+                &self.layered_snapshots,
+            ),
+            (
+                "Hybrid",
+                Box::new(|| Box::new(self.hybrid.clone())),
+                &self.hybrid_checkpoints,
+                &self.hybrid_snapshots,
+            ),
         ]
     }
+
+    fn assert_scale_divisibility() {
+        // CHECKPOINT_SCALES needs to be divisble (without remainder) by every scale of
+        // SNAPSHOT_SCALES or the benchmark will end up having fluctuations in number
+        // of accounts, etc. due to rounding. The same applies for ADDRESS_SCALES
+        // needing to be divisible (without remainder) by every scale of
+        // CHECKPOINT_SCALES.
+
+        for address_scale in ADDRESS_SCALES {
+            for checkpoint_scale in CHECKPOINT_SCALES {
+                assert!(address_scale % checkpoint_scale == 0, "all address scales must be evenly divisible by all checkpoint scales, but address scale {address_scale} is not evenly divisible by checkpoint scale {checkpoint_scale}");
+            }
+        }
+        for checkpoint_scale in CHECKPOINT_SCALES {
+            for snapshot_scale in SNAPSHOT_SCALES {
+                assert!(checkpoint_scale % snapshot_scale == 0, "all checkpoint scales must be evenly divisible by all snapshots scales, but checkpoint scale {checkpoint_scale} is not evenly divisible by snapshot scale {snapshot_scale}");
+            }
+        }
+    }
+}
+
+impl Default for RethnetStates {
+    fn default() -> Self {
+        Self::assert_scale_divisibility();
```

Comment:
Yes, I did do this in a weird way, so thank you for pushing back on it.

What I had wanted to do was a file level assert, just after the declarations, but Rust doesn't support that without a 3rd party crate. I chose to do it here rather than in `fill()` because this struct could in theory be used by some other benchmark without that benchmark using `fill()`. But I agree this isn't ideal. I took your idea and then took it a little further, using a `struct` and `impl` rather than the `mod`. Check it out in https://github.com/NomicFoundation/hardhat/pull/3921/commits/9d1de6d50964a9a650fdb5fdd63f3dc12ce2d7e5